README.md
# AlarmDecoder Arduino library for ESP32

This project provides a framework for building a network IoT appliance to monitor and control one or thousands of alarm panels using an AD2pHAT from AlarmDecoder.com

The AD2EmbeddedIot Arduino sketch example has the following features that can be selectively enabled or disabled.
- HTTP/HTTPS web server with REST API.
- MQTT/MQTTS client to connect to public or private MQTT/MQTTS broker.
- SSDP UPNP. Provides discovery capabilities with Hubitat Elevation™, SmartThings™ hub and other SSDP capable controllers.
- Socket server. Allows connections from the AlarmDecoder keypad GUI app, Micasaverde VERA, Hasbian and and other systems that support TCP/IP connections.
- AlarmDecoder protocol pass through. Connect the ESP32 USB port to a host computer to access the AD2 device directly using a serial terminal.

Using an Olimex ESP32-EVB-EA it takes 6 seconds after booting to connect to the network and start sending messages.

## Getting the code
## Configuring
## Building

## Contributors
 - Submit issues and contribute improvements on [github/nutechsoftware](https://github.com/nutechsoftware)

## Authors
 - Sean Mathews <coder@f34r.com> - Initial skeleton and R&D

## License
 - [Apache 2.0](http://www.apache.org/licenses/LICENSE-2.0)


CHANGELOG.md
#Changelog
All notible changes to this project will be documented in this file.

- 1.0 - 2020-01-15
  - Sean Mathews <coder@f34r.com> - Initial skeleton of library


src/ArduinoAlarmDecoder.h
/**
 *  @file    ArduinoAlarmDecoder.h
 *  @author  Sean Mathews <coder@f34r.com>
 *  @date    01/15/2020
 *  @version 1.0
 *
 *  @brief AlarmDecoder embedded state machine and parser
 *
 *  @copyright Copyright (C) 2020 Nu Tech Software Solutions, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
#ifndef AlarmDecoder_h
#define AlarmDecoder_h
#include <stdint.h>
#include <WString.h>
#include <map>
#include "Arduino.h"

// types and defines
typedef void (*AD2ParserCallback_msg_t)(void*);


enum AD2_PARSER_STATES {
	AD2_PARSER_RESET            = 0,
	AD2_PARSER_SCANNING_START   = 1,
	AD2_PARSER_SCANNING_EOL     = 2,
	AD2_PARSER_PROCESSING       = 3
};

// The actual max is ~90 but leave some room for future.
// WARN: use int8_t so must be <= 127 for overflow protection.
#define ALARMDECODER_MAX_MESSAGE_SIZE 120

#define BIT_ON '1'
#define BIT_OFF '0'
#define BIT_UNDEFINED '-'

// BIT/DATA OFFSETS
#define READY_BYTE          1
#define ARMED_AWAY_BYTE     2
#define ARMED_HOME_BYTE     3
#define BACKLIGHT_BYTE      4
#define PROGMODE_BYTE       5
#define BEEPMODE_BYTE       6
#define BYPASS_BYTE         7
#define ACPOWER_BYTE        8
#define CHIME_BYTE          9
#define ALARMSTICKY_BYTE   10
#define ALARM_BYTE         11
#define LOWBATTERY_BYTE    12
#define ENTRYDELAY_BYTE    13
#define FIRE_BYTE          14
#define SYSISSUE_BYTE      15
#define PERIMETERONLY_BYTE 16

#define AMASK_START        30
#define AMASK_END          38



/**
 * Utility functions/macros.
 */

#define AD2_NTOHL(x) ((((x) & 0xff000000UL) >> 24) | \
                    (((x) & 0x00ff0000UL) >>  8) | \
                    (((x) & 0x0000ff00UL) <<  8) | \
                    (((x) & 0x000000ffUL) << 24))

/**
 * Data structure for each Virtual partition state.
 *
 * https://www.alarmdecoder.com/wiki/index.php/Protocol
 *
 * The AlarmDecoder Alphanumeric Keypad Message has 4 sections.
 * 1) Bit Fields   - [00110011000000003A--]
 * 2) Numeric data - 010
 * 3) Raw Data     - [f70200010010808c18020000000000]
 *                      ^^^^^^^^ <- Address bytes
 * 4) Alpha Data   - "ARMED ***STAY** ZONE BYPASSED "
 *
 * The Partition/Address data is kept in section #3
 * bytes 2-9
 *  https://www.alarmdecoder.com/wiki/index.php/Protocol#Raw_data
 *
 * This 32 bit Little Endian word uses each bit to represent a single
 * address(Ademco) or partition(DSC) from 0 to 31.
 *
 * Ademco example
 *  0x02000100 = This message was for keypads at address 1 and 16.
 *
 * DSC example
 *  0x02000000 = This message was for partition #1
 *
 * To prevent mixing data between partitions a separate instance of this
 * storage structure will be used for each unique mask received from the
 * AlarmDecoder using the 32 bit address mask as a unique key.
 *
 * To determin the actual partition for each structure on Ademco it would be
 * necessary to have a translation between keypad address and partition.
 *
 */
class AD2VirtualPartitionState
{
	public:

  // Address mask filter for this partition
	uint32_t address_mask_filter;

  // Partition number(external lookup required for Ademco)
  uint8_t partition;

  // Calculated from section #3(Raw)
	uint8_t display_cursor_location; // 1-32
  uint8_t display_cursor_type;     // 0[OFF] 1[UNDERLINE] 2[INVERT]


  // SECTION #1 data
  //  https://www.alarmdecoder.com/wiki/index.php/Protocol#Bit_field
	bool ready;
	bool armed_away;
	bool armed_home;
	bool backlight_on;
	bool programming_mode;
	bool zone_bypassed;
	bool ac_power;
	bool chime_on;
	bool alarm_event_occurred;
	bool alarm_sounding;
	bool battery_low;
	bool entry_delay_off;
	bool fire_alarm;
	bool check_zone;
	bool perimeter_only;
	bool system_fault;
	uint8_t beeps;
	char panel_type;
  bool unused1;
  bool unused2;

	String last_alpha_message;
	String last_numeric_message;

};

typedef std::map<uint32_t, AD2VirtualPartitionState *> ad2pstates_t;

/**
 * AlarmDecoder protocol parser.
 * 1) Able to receive partial messages with subsequent calls to complete full
 *   messages.
 * 2) consume all data received and if not complete preserve for sub.
 * 2) upon receiving a full message parse and call any events that trigger.
 * 3) continue processing data
 */
class AlarmDecoderParser
{
  public:

		AlarmDecoderParser();

    // Subscribe to callbacks.
		void setCB_ON_RAW_MESSAGE(AD2ParserCallback_msg_t cb);
    void setCB_ON_ARM(AD2ParserCallback_msg_t cb);
    void setCB_ON_DISARM(AD2ParserCallback_msg_t cb);
		void setCB_ON_POWER_CHANGE(AD2ParserCallback_msg_t cb);
		void setCB_ON_READY_CHANGE(AD2ParserCallback_msg_t cb);
		void setCB_ON_ALARM(AD2ParserCallback_msg_t cb);
		void setCB_ON_ALARM_RESTORED(AD2ParserCallback_msg_t cb);
		void setCB_ON_FIRE(AD2ParserCallback_msg_t cb);
		void setCB_ON_BYPASS(AD2ParserCallback_msg_t cb);
		void setCB_ON_BOOT(AD2ParserCallback_msg_t cb);
		void setCB_ON_CONFIG_RECEIVED(AD2ParserCallback_msg_t cb);
		void setCB_ON_ZONE_FAULT(AD2ParserCallback_msg_t cb);
		void setCB_ON_ZONE_RESTORE(AD2ParserCallback_msg_t cb);
		void setCB_ON_LOW_BATTERY(AD2ParserCallback_msg_t cb);
		void setCB_ON_PANIC(AD2ParserCallback_msg_t cb);
		void setCB_ON_RELAY_CHANGED(AD2ParserCallback_msg_t cb);
		void setCB_ON_CHIME_CHANGED(AD2ParserCallback_msg_t cb);
		void setCB_ON_MESSAGE(AD2ParserCallback_msg_t cb);
		void setCB_ON_EXPANDER_MESSAGE(AD2ParserCallback_msg_t cb);
		void setCB_ON_LRR(AD2ParserCallback_msg_t cb);
		void setCB_ON_RFX(AD2ParserCallback_msg_t cb);
		void setCB_ON_SENDING_RECEIVED(AD2ParserCallback_msg_t cb);
		void setCB_ON_AUI(AD2ParserCallback_msg_t cb);
		void setCB_ON_KPM(AD2ParserCallback_msg_t cb);
		void setCB_ON_KPE(AD2ParserCallback_msg_t cb);
		void setCB_ON_CRC(AD2ParserCallback_msg_t cb);
		void setCB_ON_VER(AD2ParserCallback_msg_t cb);
		void setCB_ON_ERR(AD2ParserCallback_msg_t cb);

    // Callback functino pointers.
		AD2ParserCallback_msg_t ON_RAW_MESSAGE_CB;
    AD2ParserCallback_msg_t ON_ARM_CB;
    AD2ParserCallback_msg_t ON_DISARM_CB;
		AD2ParserCallback_msg_t ON_POWER_CHANGE_CB;
		AD2ParserCallback_msg_t ON_READY_CHANGE_CB;
		AD2ParserCallback_msg_t ON_ALARM_CB;
		AD2ParserCallback_msg_t ON_ALARM_RESTORED_CB;
		AD2ParserCallback_msg_t ON_FIRE_CB;
		AD2ParserCallback_msg_t ON_BYPASS_CB;
		AD2ParserCallback_msg_t ON_BOOT_CB;
		AD2ParserCallback_msg_t ON_CONFIG_RECEIVED_CB;
		AD2ParserCallback_msg_t ON_ZONE_FAULT_CB;
		AD2ParserCallback_msg_t ON_ZONE_RESTORE_CB;
		AD2ParserCallback_msg_t ON_LOW_BATTERY_CB;
	  AD2ParserCallback_msg_t ON_PANIC_CB;
		AD2ParserCallback_msg_t ON_RELAY_CHANGED_CB;
		AD2ParserCallback_msg_t ON_CHIME_CHANGED_CB;
		AD2ParserCallback_msg_t ON_MESSAGE_CB;
    AD2ParserCallback_msg_t ON_EXPANDER_MESSAGE_CB;
		AD2ParserCallback_msg_t ON_LRR_CB;
		AD2ParserCallback_msg_t ON_RFX_CB;
		AD2ParserCallback_msg_t ON_SENDING_RECEIVED_CB;
		AD2ParserCallback_msg_t ON_AUI_CB;
		AD2ParserCallback_msg_t ON_KPM_CB;
		AD2ParserCallback_msg_t ON_KPE_CB;
		AD2ParserCallback_msg_t ON_CRC_CB;
		AD2ParserCallback_msg_t ON_VER_CB;
		AD2ParserCallback_msg_t ON_ERR_CB;


		// Push data into state machine. Events fire if a complete message is
		// received.
		bool put(uint8_t *buf, int8_t len);

    // Reset the parser state machine.
    void reset_parser();


    void test();




	protected:
		// Track all panel states in separate class.
		ad2pstates_t AD2PStates;

	  // Parser state control starts out as AD2_PARSER_RESET.
		int AD2_Parser_State;

    // ring buffer must be < 128 bytes using signed int8 for pointer variable.
		uint8_t ring_buffer[ALARMDECODER_MAX_MESSAGE_SIZE];
		int8_t ring_qin_position, ring_qout_position;
    uint16_t ring_error_count;

};



// Forward decls.
void ad2test(void);

// Utility functions.
bool is_bit_set(int pos, const char * bitStr);

#endif


src/ArduinoAlarmDecoder.cpp
/**
 *  @file    ArduinoAlarmDecoder.cpp
 *  @author  Sean Mathews <coder@f34r.com>
 *  @date    01/15/2020
 *  @version 1.0
 *
 *  @brief AlarmDecoder embedded state machine and parser
 *
 *  @copyright Copyright (C) 2020 Nu Tech Software Solutions, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

#include "ArduinoAlarmDecoder.h"



AlarmDecoderParser::AlarmDecoderParser() {

  // clear all callback pointers.
  ON_RAW_MESSAGE_CB = 0;
  ON_MESSAGE_CB = 0;
  ON_ARM_CB = 0;
  ON_DISARM_CB = 0;
  ON_POWER_CHANGE_CB = 0;
  ON_READY_CHANGE_CB = 0;
  ON_ALARM_CB = 0;
  ON_ALARM_RESTORED_CB = 0;
  ON_FIRE_CB = 0;
  ON_BYPASS_CB = 0;
  ON_BOOT_CB = 0;
  ON_CONFIG_RECEIVED_CB = 0;
  ON_ZONE_FAULT_CB = 0;
  ON_ZONE_RESTORE_CB = 0;
  ON_LOW_BATTERY_CB = 0;
  ON_PANIC_CB = 0;
  ON_RELAY_CHANGED_CB = 0;
  ON_CHIME_CHANGED_CB = 0;
  ON_MESSAGE_CB = 0;
  ON_EXPANDER_MESSAGE_CB = 0;
  ON_LRR_CB = 0;
  ON_RFX_CB = 0;
  ON_SENDING_RECEIVED_CB = 0;
  ON_AUI_CB = 0;
  ON_KPM_CB = 0;
  ON_KPE_CB = 0;
  ON_CRC_CB = 0;
  ON_VER_CB = 0;
  ON_ERR_CB = 0;

  // Reset the parser on init.
  reset_parser();

}

void AlarmDecoderParser::reset_parser() {
  // Initialize parser state machine state.
  AD2_Parser_State = AD2_PARSER_RESET;
}

/**
 * Consume bytes from an AlarmDecoder stream into a small ring buffer
 * for processing.
 *
 * 1) Parse all of the data firing off events upon parsing a full message.
 *   Continue parsing data until all is consumed.
 */
bool AlarmDecoderParser::put(uint8_t *buff, int8_t len) {

  // All AlarmDecoder messages are '\n' terminated.
  // "!boot.....done" is the only state exists that needs notification
  // before the terminator in order to enter the boot loader.
  // If KPM config bit is set then all messages will start with '!'
  // If KPM config bit is not set(the default) then standard keypad state
  // messages start with '['.

  int8_t bytes_left = len;
  uint8_t *bp = buff;

  // Sanity check.
  if (len<=0) {
     return false;
  }

  // Consume all the bytes.
  while (bytes_left>0) {

    uint8_t ch;

    // Update state machine.
    switch (AD2_Parser_State) {

      // Reset the ring buffer state.
      case AD2_PARSER_RESET:

        ring_qin_position = 0;
        ring_qout_position = 0;

        // Initialize AlarmDecoder stream parsing ring buffer memory to 0s.
        memset(ring_buffer, 0, sizeof(ring_buffer));
        AD2_Parser_State = AD2_PARSER_SCANNING_START;
        break;

      // Wait for ALPHA/NUMERIC after an EOL
      case AD2_PARSER_SCANNING_START:
        // store local.
        ch = *bp;

        // start scanning for EOL as soon as we have a printable character.
        if (ch >31 && ch <127) {
          AD2_Parser_State = AD2_PARSER_SCANNING_EOL;
        } else {
          // Dump the byte.
          // Update remaining bytes counter and move ptr.
          bp++;
          bytes_left--;
        }
        break;

      // Consume bytes looking for terminator.
      case AD2_PARSER_SCANNING_EOL:
        // store local.
        ch = *bp;

        // Protection from external leaks or faults. Very unlikely to happen.
        // Do not consume the byte and try again after a reset.
        if (ring_qin_position > ALARMDECODER_MAX_MESSAGE_SIZE) {

          // Next state is RESET.
          AD2_Parser_State = AD2_PARSER_RESET;

          // Done.
          break;
        }

        // Protect from corrupt data skip and reset.
        // All bytes must be CR/LF or printable characters only.
        if (ch != '\r' && ch != '\n') {
          if (ch <32 || ch >126) {

            // Update remaining bytes counter and move ptr.
            bp++;
            bytes_left--;

            // Next state is RESET.
            AD2_Parser_State = AD2_PARSER_RESET;

            // Done.
            break;
          }
        }

        // Update remaining bytes counter and move ptr.
        bp++;
        bytes_left--;

        // Process full messages on CR or LF
        if ( ch == '\n' || ch == '\r') {

          // Next wait for start of next message
          AD2_Parser_State = AD2_PARSER_SCANNING_START;

          // FIXME: Make it better faster less cpu mem etc. String :(
          // Consume all of the data for this message from the ring buffer.
          String msg = "";
          while (ring_qout_position != ring_qin_position) {
            msg += (char)ring_buffer[ring_qout_position];
            // Reset to 0 at the end of the ring.
            if ( ++ring_qout_position >= ALARMDECODER_MAX_MESSAGE_SIZE ) {
              ring_qout_position = 0;
            }
          }

          // call ON_RAW_MESSAGE callback if enabled.
          if (ON_RAW_MESSAGE_CB) {
            ON_RAW_MESSAGE_CB((void*)&msg);
          }

          // Detect message type or error.
          // 1) Starts with !
          //     !boot, !EXP, !REL, etc, etc.
          // 2) Starts with [
          //    Standard status message.
          // All other cases are invalid
          //
          if (msg[0] == '!') {
            if (msg.startsWith("!LRR:")) {
              // call ON_LRR callback if enabled.
              if (ON_LRR_CB) {
                ON_LRR_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!REL:") || msg.startsWith("!EXP:")) {
              // call ON_EXPANDER_MESSAGE callback if enabled.
              if (ON_EXPANDER_MESSAGE_CB) {
                ON_EXPANDER_MESSAGE_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!RFX:")) {
              // call ON_RFX callback if enabled.
              if (ON_RFX_CB) {
                ON_RFX_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!AUI:")) {
              // call ON_AUI callback if enabled.
              if (ON_AUI_CB) {
                ON_AUI_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!KPM:")) {
              // call ON_KPM callback if enabled.
              if (ON_KPM_CB) {
                ON_KPM_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!KPE:")) {
              // call ON_KPE callback if enabled.
              if (ON_KPE_CB) {
                ON_KPE_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!CRC:")) {
              // call ON_CRC callback if enabled.
              if (ON_CRC_CB) {
                ON_CRC_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!VER:")) {
              // Parse the version string.
              // call ON_VER callback if enabled.
              if (ON_VER_CB) {
                ON_VER_CB((void*)&msg);
              }
            } else
            if (msg.startsWith("!ERR:")) {
              // call ON_ERR callback if enabled.
              if (ON_ERR_CB) {
                ON_ERR_CB((void*)&msg);
              }
            }
          } else {
            // http://www.alarmdecoder.com/wiki/index.php/Protocol#Keypad
            if (msg[0] == '[') {

              // Excessive sanity check. Test a few static characters.
              // Length should be 94 bytes and end with ".
              // [00110011000000003A--],010,[f70700000010808c18020000000000],"ARMED ***STAY** ZONE BYPASSED "
              if (msg.length() == 94 && msg[93]=='"' && msg[22] == ',') {

                // First extract the 32 bit address mask from section #3
                // to use it as a storage key for the state.
                uint32_t amask = strtol( msg.substring(AMASK_START,
                                        AMASK_END).c_str(), NULL, 16);

                amask = AD2_NTOHL(amask);
                // Ademco/DSC: MASK 00000000 = System
                // Ademco 00000001 is keypad address 0
                // Ademco 00000002 is keypad address 1
                // DSC    00000002 is partition 1
                // Ademco 40000000 is keypad address 30

                // Create or return a pointer to our partition storage class.
                AD2VirtualPartitionState *ad2ps = NULL;

                // look for an exact match.
                if ( AD2PStates.find(amask) == AD2PStates.end()) {

                  // Not found create or find a mask that has at least
                  // one bit in common and update its mask to include the new
                  // bits. System partition mask is 0 and will skip this logic.
                  if (amask) {
                      uint32_t foundkey = amask;
                      for (auto const& x : AD2PStates)
                      {
                        // Mask has matching bits use it instead.
                        if (x.first & amask) {
                          ad2ps = AD2PStates[x.first];
                          foundkey = x.first;
                          break;
                        }
                      }

                      // Not necessary but seems reasonable.
                      // Update key to include new mask if it changed.
                      if (foundkey != amask) {
                        // remove the old map entry.
                        AD2PStates.erase(foundkey);
                        // Add new one with mask of original + new.
                        AD2PStates[amask | foundkey] = ad2ps;
                      }
                  }

                  // Did not find entry. Make new.
                  if (!ad2ps) {
                      ad2ps = AD2PStates[amask] = new AD2VirtualPartitionState;
                      ad2ps->partition = AD2PStates.size();
                  }

                } else {
                  // Found. Grab the state class ptr.
                  ad2ps = AD2PStates[amask];
                }


                // Update the partition state based upon the new status message.

                // Extract state bits from section #1
                ad2ps->ready      = is_bit_set( READY_BYTE, msg.c_str());
                ad2ps->armed_away = is_bit_set( ARMED_AWAY_BYTE, msg.c_str());
                ad2ps->armed_home = is_bit_set( ARMED_HOME_BYTE, msg.c_str());
                ad2ps->zone_bypassed = is_bit_set( BYPASS_BYTE, msg.c_str());

                // Extract the numberic value from section #2

                // Extract the Alpha message from section #4.

                // Extract the cursor location and type from section #3

                // FIXME: debugging / testing
                Serial.print("!DBG: SIZE(");
                Serial.print(AD2PStates.size());
                Serial.print(") PID(");
                Serial.print(ad2ps->partition);
                Serial.print(") MASK(");
                Serial.print(amask,BIN);
                Serial.print(") Ready(");
                Serial.print(ad2ps->ready);
                Serial.print(") Armed Away(");
                Serial.print(ad2ps->armed_away);
                Serial.print(") Armed Home(");
                Serial.print(ad2ps->armed_home);
                Serial.print(") Bypassed(");
                Serial.print(ad2ps->zone_bypassed);
                Serial.println(")");

                // call ON_MESSAGE callback if enabled.
                if (ON_MESSAGE_CB) {
                  ON_MESSAGE_CB((void*)&msg);
                }
              }
            } else {
              //FIXME: Error
              Serial.println("!ERR: BAD PROTOCOL PREFIX.");
            }
          }


          // Do not save EOL into the ring. We are done for now.
          break;
        }

        // Still receiving a message.
        // Save this byte to our ring buffer and parse and keep waiting for EOL.
        ring_buffer[ring_qin_position] = ch;

        // Reset to 0 at the end of the ring
        if ( ++ring_qin_position >= ALARMDECODER_MAX_MESSAGE_SIZE ) {
          ring_qin_position = 0;
        }

        // If next qin address is qout then loose a byte and report error.
        if (ring_qin_position == ring_qout_position) {
          ring_error_count++;
          if ( ++ring_qout_position >= ALARMDECODER_MAX_MESSAGE_SIZE ) {
            ring_qout_position = 0;
          }
        }

        break;

      case AD2_PARSER_PROCESSING:
        // FIXME: TBD
        break;

      default:
        // FIXME: TBD
        break;
    }
  }

  return true;
}

/**
 * FIXME: test code
 */
void AlarmDecoderParser::test() {
  for (int x = 0; x < 10000; x++) {
    AD2PStates[1] = new AD2VirtualPartitionState;
    AD2PStates[1]->ready = false;
    delete AD2PStates[1];
  }
}

/**
 * setCB_ON_RAW_MESSAGE
 */
void AlarmDecoderParser::setCB_ON_RAW_MESSAGE(AD2ParserCallback_msg_t cb) {
  ON_RAW_MESSAGE_CB = cb;
}

/**
 * setCB_ON_ARM
 */
void AlarmDecoderParser::setCB_ON_ARM(AD2ParserCallback_msg_t cb) {
  ON_ARM_CB = cb;
}

/**
 * setCB_ON_DISARM
 */
void AlarmDecoderParser::setCB_ON_DISARM(AD2ParserCallback_msg_t cb) {
  ON_DISARM_CB = cb;
}

/**
 * setCB_POWER_CHANGE
 */
void AlarmDecoderParser::setCB_ON_POWER_CHANGE(AD2ParserCallback_msg_t cb) {
  ON_POWER_CHANGE_CB = cb;
}

/**
 * setCB_ON_READY_CHANGE
 */
void AlarmDecoderParser::setCB_ON_READY_CHANGE(AD2ParserCallback_msg_t cb) {
  ON_READY_CHANGE_CB = cb;
}

/**
 * setCB_ON_ALARM
 */
void AlarmDecoderParser::setCB_ON_ALARM(AD2ParserCallback_msg_t cb) {
  ON_ALARM_CB = cb;
}

/**
 * setCB_ON_ALARM_RESTORED
 */
void AlarmDecoderParser::setCB_ON_ALARM_RESTORED(AD2ParserCallback_msg_t cb) {
  ON_ALARM_RESTORED_CB = cb;
}

/**
 * setCB_ON_FIRE
 */
void AlarmDecoderParser::setCB_ON_FIRE(AD2ParserCallback_msg_t cb) {
  ON_FIRE_CB = cb;
}

/**
 * setCB_ON_BYPASS
 */
void AlarmDecoderParser::setCB_ON_BYPASS(AD2ParserCallback_msg_t cb) {
  ON_BYPASS_CB = cb;
}

/**
 * setCB_ON_BOOT
 */
void AlarmDecoderParser::setCB_ON_BOOT(AD2ParserCallback_msg_t cb) {
  ON_BOOT_CB = cb;
}

/**
 * setCB_ON_CONFIG_RECEIVED
 */
void AlarmDecoderParser::setCB_ON_CONFIG_RECEIVED(AD2ParserCallback_msg_t cb) {
  ON_CONFIG_RECEIVED_CB = cb;
}


/**
 * setCB_ON_ZONE_FAULT
 */
void AlarmDecoderParser::setCB_ON_ZONE_FAULT(AD2ParserCallback_msg_t cb) {
  ON_ZONE_FAULT_CB = cb;
}

/**
 * setCB_ON_ZONE_RESTORE
 */
void AlarmDecoderParser::setCB_ON_ZONE_RESTORE(AD2ParserCallback_msg_t cb) {
  ON_ZONE_RESTORE_CB = cb;
}

/**
 * setCB_ON_LOW_BATTERY
 */
void AlarmDecoderParser::setCB_ON_LOW_BATTERY(AD2ParserCallback_msg_t cb) {
  ON_LOW_BATTERY_CB = cb;
}

/**
 * setCB_ON_PANIC
 */
void AlarmDecoderParser::setCB_ON_PANIC(AD2ParserCallback_msg_t cb) {
  ON_PANIC_CB = cb;
}

/**
 * setCB_ON_RELAY_CHANGED
 */
void AlarmDecoderParser::setCB_ON_RELAY_CHANGED(AD2ParserCallback_msg_t cb) {
  ON_RELAY_CHANGED_CB = cb;
}

/**
 * setCB_ON_CHIME_CHANGED
 */
void AlarmDecoderParser::setCB_ON_CHIME_CHANGED(AD2ParserCallback_msg_t cb) {
  ON_CHIME_CHANGED_CB = cb;
}

/**
 * setCB_ON_MESSAGE
 */
void AlarmDecoderParser::setCB_ON_MESSAGE(AD2ParserCallback_msg_t cb) {
  ON_MESSAGE_CB = cb;
}

/**
 * setCB_ON_EXPANDER_MESSAGE
 */
void AlarmDecoderParser::setCB_ON_EXPANDER_MESSAGE(AD2ParserCallback_msg_t cb) {
  ON_EXPANDER_MESSAGE_CB = cb;
}

/**
 * setCB_ON_LRR
 */
void AlarmDecoderParser::setCB_ON_LRR(AD2ParserCallback_msg_t cb) {
  ON_LRR_CB = cb;
}

/**
 * setCB_ON_RFX
 */
void AlarmDecoderParser::setCB_ON_RFX(AD2ParserCallback_msg_t cb) {
  ON_RFX_CB = cb;
}

/**
 * setCB_ON_SENDING_RECEIVED
 */
void AlarmDecoderParser::setCB_ON_SENDING_RECEIVED(AD2ParserCallback_msg_t cb) {
  ON_SENDING_RECEIVED_CB = cb;
}

/**
 * setCB_ON_AUI
 */
void AlarmDecoderParser::setCB_ON_AUI(AD2ParserCallback_msg_t cb) {
  ON_AUI_CB = cb;
}

/**
 * setCB_ON_KPM
 */
void AlarmDecoderParser::setCB_ON_KPM(AD2ParserCallback_msg_t cb) {
  ON_KPM_CB = cb;
}

/**
 * setCB_ON_KPE
 */
void AlarmDecoderParser::setCB_ON_KPE(AD2ParserCallback_msg_t cb) {
  ON_KPE_CB = cb;
}

/**
 * setCB_ON_CRC
 */
void AlarmDecoderParser::setCB_ON_CRC(AD2ParserCallback_msg_t cb) {
  ON_CRC_CB = cb;
}

/**
 * setCB_ON_VER
 */
void AlarmDecoderParser::setCB_ON_VER(AD2ParserCallback_msg_t cb) {
  ON_VER_CB = cb;
}

/**
 * setCB_ON_ERR
 */
void AlarmDecoderParser::setCB_ON_ERR(AD2ParserCallback_msg_t cb) {
  ON_ERR_CB = cb;
}



/**
* function: is_bit_set
* check to see if bits are enabled in status field
*
* in: int
* description: position in buffer to check
*
* in: char *
* description: complete status field buffer
 *
*/
bool is_bit_set(int pos, const char * bitStr)
{
      int set;
      set = 0;

      if( bitStr[pos] == '0' || bitStr[pos] == BIT_UNDEFINED)
              set = 0;
      if( bitStr[pos] == BIT_ON )
              set = 1;

      return set;
}
